font/cache.go

```diff
@@ -0,0 +1,38 @@
+package font
+
+type glyphExtents struct {
+	valid   bool
+	extents GlyphExtents
+}
+
+type extentsCache []glyphExtents
+
+func (ec extentsCache) get(gid GID) (GlyphExtents, bool) {
+	if int(gid) >= len(ec) {
+		return GlyphExtents{}, false
+	}
+	ge := ec[gid]
+	return ge.extents, ge.valid
+}
+
+func (ec extentsCache) set(gid GID, extents GlyphExtents) {
+	ec[gid].valid = true
+	ec[gid].extents = extents
+}
```

Comment:
`get` has a range check to ensures that `gid` is in bounds, but `set` does not. Seems like we're risking OOB access here.

Also, I would expect that GIDs are accessed somewhat sparsely and thus we'd end up with a very sparsely-populated slice. It might be more memory efficient to use a different data structure or to store elements sorted by GID that we can resolve with a binary search (though inserting would be expensive then).

Reply:
> `set` does not. Seems like we're risking OOB access here.

The idea is that we only use `set` if `glyphExtentsRaw` returns true, which will never happen if the GID is out of bounds.

> Also, I would expect that GIDs are accessed somewhat sparsely and thus we'd end up with a very sparsely-populated slice. It might be more memory efficient to use a different data structure or to store elements sorted by GID that we can resolve with a binary search (though inserting would be expensive then).

Your right in theory. In practice, I'm not sure the additional memory usage is very significant, compared to the bytes needed to encode glyph segments, for instance.
Since it is entirely internal, perhaps such an optimization could be postponed to a subsequent PR ?

Reply:
I've updated `TestFontLoadHeapSize` to backup my point. It  yields the following measurements : 

arabic:amiri regular On disk: 540 KB, font memory: 2248 KB, face memory 136 KB
latin:go regular On disk: 145 KB, font memory: 344 KB, face memory 14 KB
latin:roboto regular On disk: 164 KB, font memory: 454 KB, face memory 26 KB
arabic:noto sans arabic regular On disk: 747 KB, font memory: 3377 KB, face memory 31 KB
free serif regular On disk: 1647 KB, font memory: 5588 KB, face memory 176 KB

The additional memory used by the cache is, in the worst case, around 5% of the total memory.

Reply:
Alright, I've added a bounds check in `set`